solarforecastarbiter/metrics/probabilistic.py

```diff
@@ -0,0 +1,272 @@
+
+    where n is the number of forecasts, f_i is the forecasted probability of
+    event i, and o_i is the observed event indicator (o_i=0: event did not
+    occur, o_i=1: event occured).
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    bs : float
+        The Brier Score [unitless], bounded between 0 and 1, where values
+        closer to 0 indicate better forecast performance and values closer to 1
+        indicate worse performance.
+
+    Notes
+    -----
+    The Brier Score implemented in this function is for binary outcomes only,
+    rather than the more general (but less commonly used) categorical version.
+
+    """
+
+    # event: 0=did not happen, 1=did happen
+    o = np.where(obs <= fx, 1.0, 0.0)
+
+    # forecast probabilities [unitless]
+    f = fx_prob / 100.0
+
+    bs = np.mean((f - o) ** 2)
+    return bs
+
+
+def brier_skill_score(fx, fx_prob, ref, ref_prob, obs):
+    """Brier Skill Score (BSS).
+
+        BSS = 1 - BS_f / BS_ref
+
+    where BS_fx is the Brier Score of the evaluated forecast and BS_ref is the
+    Brier Score of a reference forecast.
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    ref : (n,) array_like
+        Reference forecast (physical units).
+    ref_prob : (n,) array_like
+        Probability [%] associated with the reference forecast.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    Returns
+    -------
+    skill : float
+        The Brier Skill Score [unitless].
+
+    """
+    bs_fx = brier_score(fx, fx_prob, obs)
+    bs_ref = brier_score(ref, ref_prob, obs)
+    skill = 1.0 - bs_fx / bs_ref
+    return skill
+
+
+def brier_decomposition(fx, fx_prob, obs):
+    """The 3-component decomposition of the Brier Score.
+
+        BS = REL - RES + UNC
+
+    where REL is the reliability, RES is the resolution and UNC is the
+    uncertatinty.
+
+    Parameters
+    ----------
+    fx : (n,) array_like
+        Forecasts (physical units).
+    fx_prob : (n,) array_like
+        Probability [%] associated with the forecasts.
+    obs : (n,) array_like
+        Observations (physical unit).
+
+    """
+
+    # event: 0=did not happen, 1=did happen
+    o = np.where(obs <= fx, 1.0, 0.0)
+
+    # forecast probabilities [unitless]
+    f = fx_prob / 100.0
+
+    # get unique forecast probabilities by binning
+    if len(f) >= 1000:
+        n_decimals = 2  # bin by hundredths (0.01, 0.02, etc.)
+    else:
+        n_decimals = 1  # bin by tenths (0.1, 0.2, etc.)
+
+    f = np.around(f, decimals=n_decimals)
+
+    # reliability
+    rel = 0.0
+    for f_i, N_i in np.nditer(np.unique(f, return_counts=True)):
```

Comment:
I'm torn here. 
1. I would often try to avoid looping over arrays, but in this case there are at most 100 iterations and I can't think of how to do this clearly without lopping. 
2. extra work is done when only calling reliability, resolution, or uncertainty. Again, the code would be more efficient to refactor, but harder to read later.

Overall, I would leave it as is but comment that it's done this way for ease of understanding.

Reply:
Although, maybe just combine the reliability and resolution loops (with tweaked variable names) and avoid looping twice.

Reply:
> Although, maybe just combine the reliability and resolution loops (with tweaked variable names) and avoid looping twice.

this was my (unclear) expectation when I suggested earlier that the two functions could be combined

Reply:
One thing I forgot to do is combine the `rel` and `res` loops (since they re-use the same variables), which will technically improve the efficiency (which I'll do now). But I definitely agree with your points.

Reply:
Somehow your comments didn't appear when I made my comment. Anyways, I've now combined the two loops and I'm working on a note in the docstring about why we're keeping a loop (rather than a vectorized implementation, which may be technically faster, but is likely less clear to most).

Reply:
I added a **Notes** section to the `brier_decomposition()` docstring that provides context for the choice of iterating over the unique forecasts (rather than a vectorized implementation). I think it covers the main points, but feel free to suggest revisions.